lwcomponent/component.go

```diff
@@ -0,0 +1,395 @@
+			return s, err
+		}
+
+		return s, s.WriteState()
+	}
+	return nil, errors.New("invalid api client")
+}
+
+// LocalState loads the state from the local storage ("Dir()/state")
+func LocalState() (*State, error) {
+	state := new(State)
+	componentsFile, err := Dir()
+	if err != nil {
+		return state, err
+	}
+
+	stateFile := filepath.Join(componentsFile, "state")
+
+	stateBytes, err := ioutil.ReadFile(stateFile)
+	if err != nil {
+		return state, err
+	}
+	err = json.Unmarshal(stateBytes, state)
+	return state, err
+}
+
+// GetComponent returns the pointer of a component, if the component does not
+// exit, this function will return `nil`
+func (s State) GetComponent(name string) *Component {
+	for i := range s.Components {
+		if s.Components[i].Name == name {
+			return &s.Components[i]
+		}
+	}
+	return nil
+}
+
+// WriteState stores the components state to disk
+func (s State) WriteState() error {
+	dir, err := Dir()
+	if err != nil {
+		return err
+	}
+
+	if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+		return err
+	}
+
+	stateFile := filepath.Join(dir, "state")
+	buf := new(bytes.Buffer)
+	if err := json.NewEncoder(buf).Encode(s); err != nil {
+		return err
+	}
+
+	if err := ioutil.WriteFile(stateFile, buf.Bytes(), 0644); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Dir returns the directory where the components will be stored
+func Dir() (string, error) {
+	cacheDir, err := cache.CacheDir()
+	if err != nil {
+		return "", errors.Wrap(err, "unable to locate components directory")
+	}
+	return filepath.Join(cacheDir, "components"), nil
+}
+
+func (s State) Install(name string) error {
+	component := s.GetComponent(name)
+	if component == nil {
+		return errors.New("component not found")
+	}
+
+	rPath, err := component.RootPath()
+	if err != nil {
+		return err
+	}
+
+	// @afiune verify if component is in latest
+
+	// @afiune install
+	if err := os.MkdirAll(rPath, os.ModePerm); err != nil {
+		return err
+	}
+
+	artifact, found := component.ArtifactForRunningHost()
+	if !found {
+		return errors.New("unsupported platform")
+	}
+
+	path, err := component.Path()
+	// it is ok if the component is not found yet
+	if err != nil && !IsNotFound(err) {
+		return err
+	}
+
+	err = downloadFile(path, artifact.URL)
+	if err != nil {
+		return errors.Wrap(err, "unable to download component artifact")
+	}
+
+	// @afiune check 1) cross-platform and 2) correct permissions
+	// if the file has permissions already, can we avoid this?
+	if component.IsExecutable() {
+		if err := os.Chmod(path, 0744); err != nil {
+			return errors.Wrap(err, "unable to make component executable")
+		}
+	}
+
+	if err := component.WriteSignature([]byte(artifact.Signature)); err != nil {
+		return err
+	}
+
+	if err := component.WriteVersion(); err != nil {
+		return err
+	}
+
+	// verify component
+	if err := component.isVerified(); err != nil {
+		// @afiune notify and remove installed component
+		defer os.RemoveAll(rPath)
+		return err
+	}
+
+	return nil
+}
```

Comment:
For all comments in `Install()` function. https://lacework.atlassian.net/browse/ALLY-1029